packages/flutter_tools/test/general.shard/vmservice_test.dart

```diff
@@ -280,8 +281,127 @@ void main() {
+        const FakeRequest(method: 'streamListen', id: '1', params: <String, Object>{
+          'streamId': 'Isolate'
+        }),
+        const FakeRequest(method: kRunInViewMethod, id: '2', params: <String, Object>{
+          'viewId': '1234',
+          'mainScript': 'main.dart',
+          'assetDirectory': 'flutter_assets/',
+        }),
+        FakeStreamResponse(
+          streamId: 'Isolate',
+          event: vm_service.Event(
+            kind: vm_service.EventKind.kIsolateRunnable,
+            timestamp: 1,
+          )
+        ),
+      ]
+    );
+
+    await fakeVmServiceHost.vmService.runInView(
+      viewId: '1234',
+      main: Uri.file('main.dart'),
+      assetsDirectory: Uri.file('flutter_assets/'),
+    );
+    expect(fakeVmServiceHost.hasRemainingExpectations, false);
+  });
+}
+
+class FakeVmServiceHost {
+  FakeVmServiceHost({
+    @required List<VmServiceExpectation> requests,
+  }) : _requests = requests {
+    _vmService = vm_service.VmService(
+      _input.stream,
+      _output.add,
+    );
+    _applyStreamListen();
+    _output.stream.listen((String data) {
+      final Map<String, Object> request = json.decode(data) as Map<String, Object>;
+      if (_requests.isEmpty) {
+        throw Exception('Unexpected request: $request');
+      }
+      final FakeRequest fakeRequest = _requests.removeAt(0) as FakeRequest;
+      expect(fakeRequest, isA<FakeRequest>()
+        .having((FakeRequest request) => request.method, 'method', request['method'])
+        .having((FakeRequest request) => request.id, 'id', request['id'])
+        .having((FakeRequest request) => request.params, 'params', request['params'])
+      );
+      _input.add(json.encode(<String, Object>{
+        'jsonrpc': '2.0',
+        'id': fakeRequest.id,
+        'result': fakeRequest.jsonResponse ?? <String, Object>{'type': 'Success'},
+      }));
+      _applyStreamListen();
+    });
+  }
+
+  final List<VmServiceExpectation> _requests;
+  final StreamController<String> _input = StreamController<String>();
+  final StreamController<String> _output = StreamController<String>();
+
+  vm_service.VmService get vmService => _vmService;
+  vm_service.VmService _vmService;
+
+  bool get hasRemainingExpectations => _requests.isNotEmpty;
+
+  // remove FakeStreamResponse objects from _requests until it is empty
+  // or until we hit a FakeRequest
+  void _applyStreamListen() {
+    while (_requests.isNotEmpty && !_requests.first.isRequest) {
+      final FakeStreamResponse response = _requests.removeAt(0) as FakeStreamResponse;
+      _input.add(json.encode(<String, Object>{
+        'jsonrpc': '2.0',
+        'method': 'streamNotify',
+        'params': <String, Object>{
+          'streamId': response.streamId,
+          'event': response.event.toJson(),
+        },
+      }));
+    }
+  }
+}
+
+abstract class VmServiceExpectation {
+  bool get isRequest;
 }
 
+class FakeRequest implements VmServiceExpectation {
```

Comment:
> I'm not sure about the naming here. FakeRequest doesn't say much, but FakeVmServiceRequest was just too verbose

I like FakeVmServiceRequest better, it's not THAT long a name.

Reply:
Done!